File: 2807-insert-greatest-common-divisors-in-linked-list/2807-insert-greatest-common-divisors-in-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def insertGreatestCommonDivisors(self, head: Optional[ListNode]) -> Optional[ListNode]:
        
        if not head.next:
            return head
        
        def gcd(n1,n2):
            
            while n2:
                n1,n2 = n2, n1 % n2
            return abs(n1)
        
        
        curr = head
        temp = head.next
        
        while temp:
            
            new = ListNode(gcd(curr.val,temp.val))
            
            curr.next = new
            new.next = temp
            curr= temp
            temp = temp.next
        
        return head
            
            
            
            
            
            
            
            
        

File: 2807-insert-greatest-common-divisors-in-linked-list/2807-insert-greatest-common-divisors-in-linked-list.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* insertGreatestCommonDivisors(struct ListNode* head){
    
    
    if (head-> next == NULL){
        return head;
    }
    
    int gcd(int a,int b){
        
        while (b){
            
            int r = a;
            a = b;
            b = r%b;
                
        }
        return a;
        
    }


    struct ListNode *h = head;
    while(head->next != NULL) {
            int r = gcd(head->val, head->next->val);
            struct ListNode* newnode = (struct ListNode*)malloc(sizeof(struct ListNode));
            newnode->val = r;
            struct ListNode *temp = head->next;
            head->next = newnode;
            newnode->next = temp;
            head = newnode->next;
        }
    return h;

}